python-checks/src/main/java/org/sonar/python/checks/tests/AssertOnDissimilarTypesCheck.java

```diff
@@ -0,0 +1,166 @@
+      if (!(qualifiedExpression.qualifier().is(Tree.Kind.NAME) && ((Name) qualifiedExpression.qualifier()).name().equals("self"))) {
+        return;
+      }
+
+      checkArguments(ctx, callExpression, qualifiedExpression);
+    });
+  }
+
+  private static void checkArguments(SubscriptionContext ctx, CallExpression callExpression, QualifiedExpression qualifiedExpression) {
+    boolean isAnAssertIdentity = assertToCheckIdentity.contains(qualifiedExpression.name().name());
+    boolean isAnAssertEquality = assertToCheckEquality.contains(qualifiedExpression.name().name());
+    boolean shouldRaise = false;
+
+    if (!isAnAssertEquality && !isAnAssertIdentity) {
+      return;
+    }
+
+    ArgList args = callExpression.argumentList();
+    if (args == null) {
+      return;
+    }
+
+    List<Argument> arguments = args.arguments();
+    if (!(arguments.size() >= 2 && arguments.get(0).is(Tree.Kind.REGULAR_ARGUMENT) && arguments.get(1).is(Tree.Kind.REGULAR_ARGUMENT))) {
+      return;
+    }
+
+    Expression left = ((RegularArgument) arguments.get(0)).expression();
+    Expression right = ((RegularArgument) arguments.get(1)).expression();
+    if (!checkArgumentsCantBeIdentical(left, right)) {
+      return;
+    } else if (isAnAssertIdentity) {
+      shouldRaise = true;
+    }
+    if (isAnAssertEquality && checkArgumentsCantBeEqual(left, right)) {
+      shouldRaise = true;
+    }
+
+    if (shouldRaise) {
+      PreciseIssue issue = ctx.addIssue(args, message(left, right));
+      getLastAssignment(left).ifPresent(assign -> issue.secondary(assign, String.format(MESSAGE_SECONDARY, ((Name)left).name())));
+      getLastAssignment(right).ifPresent(assign -> issue.secondary(assign, String.format(MESSAGE_SECONDARY, ((Name)right).name())));
+    }
+  }
+
+  private static String message(Expression left, Expression right) {
+    String leftTypeName = InferredTypes.typeName(left.type());
+    String rightTypeName = InferredTypes.typeName(right.type());
+    String message = MESSAGE;
+    if (leftTypeName != null && rightTypeName != null) {
+      message += " (" + leftTypeName + " and " + rightTypeName + ")";
+    }
+    message += ".";
+    return message;
+  }
+
+  private static Optional<AssignmentStatement> getLastAssignment(Expression expr) {
+    if (!expr.is(Tree.Kind.NAME)) {
+      return Optional.empty();
+    }
+
+    Symbol symbol = ((Name) expr).symbol();
+    if (symbol == null) {
+      return Optional.empty();
+    }
+
+    Usage lastAssignment = symbol.usages().stream()
+      .takeWhile(usage -> usage != ((Name) expr).usage())
+      .filter(usage -> usage.kind() == Usage.Kind.ASSIGNMENT_LHS)
+      .reduce((first, second) -> second)
+      .orElse(null);
+
+    return Optional.ofNullable(lastAssignment)
+      .map(assignment -> (AssignmentStatement) TreeUtils.firstAncestorOfKind(assignment.tree(), Tree.Kind.ASSIGNMENT_STMT));
+  }
+
+  private static boolean checkArgumentsCantBeIdentical(Expression left, Expression right) {
```

Comment:
I got the same thought, I noticed that we are calling this method all the time with inverting the boolean result, so I was tempted to invert the method name/result directly, but then I noticed its "brother" method checkArgumentsCantBeEqual does not have the same case, so it would have mean to either break consistency between them or to choose one which result will be always inverted.

From my perspective, I would put both of them as "CanBeEqual" instead of their current behavior/name "CantBeEqual", to avoid encountering the double negation "!CantBeEqual" which seems less readable for me. I replaced to this way.

Agree with your last paragraph, I switched to "canArgumentsBeXXX" since "checkXXX" method is usually for method raising issues.